### ISSUE TITLE
ActiveRecord Preloading and accented characters

### ISSUE THREAD

**Issue by brandonpancost** [rails/rails#24364]
### Steps to reproduce
Create a set of two tables where one has many of the other, and the key is some text field like below.
```
class` First < ActiveRecord::Base
  has_many :second, primary_key: 'text_field', foreign_key: 'text_field'
end
class Second < ActiveRecord::Base
  belongs_to :first, :foreign_key=> 'text_field'
end
```
The text key in one table should be a string with an accented character, such as 'el_niño', and the other should be the same text without the accent such as 'el_nino'. Use a database such as MySQL where accented characters are considered to be equivalent to non-accented characters.

Attempt to use preloading like below:
```
first = First.preload(:second).where(text_field: 'el_nino')
first.each do |row|
  #do stuff
end
```

### Expected behavior
There should either throw a useful error message indicating that there is a mismatch between the values, or it should treat the keys as equivalent (or have some option to allow this behavior).

### Actual behavior
An error 'undefined method `each' for nil:NilClass' is thrown for the example above and the stack trace gives some indication there is an issue with the preloader.

```
activerecord (4.1.14.1) lib/active_record/associations/preloader/association.rb:93:in `block in associated_records_by_owner'
activerecord (4.1.14.1) lib/active_record/associations/preloader/association.rb:92:in `each'
activerecord (4.1.14.1) lib/active_record/associations/preloader/association.rb:92:in `associated_records_by_owner'
activerecord (4.1.14.1) lib/active_record/associations/preloader/collection_association.rb:13:in `preload'
activerecord (4.1.14.1) lib/active_record/associations/preloader/association.rb:20:in `run'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:137:in `block (2 levels) in preloaders_for_one'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:135:in `each'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:135:in `map'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:135:in `block in preloaders_for_one'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:134:in `each'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:134:in `flat_map'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:134:in `preloaders_for_one'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:106:in `preloaders_on'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:94:in `block in preload'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:93:in `each'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:93:in `flat_map'
activerecord (4.1.14.1) lib/active_record/associations/preloader.rb:93:in `preload'
activerecord (4.1.14.1) lib/active_record/relation.rb:617:in `block in exec_queries'
activerecord (4.1.14.1) lib/active_record/relation.rb:616:in `each'
activerecord (4.1.14.1) lib/active_record/relation.rb:616:in `exec_queries'
activerecord (4.1.14.1) lib/active_record/relation.rb:493:in `load'
activerecord (4.1.14.1) lib/active_record/relation.rb:238:in `to_a'
will_paginate (3.1.0) lib/will_paginate/active_record.rb:126:in `block in to_a'
will_paginate (3.1.0) lib/will_paginate/collection.rb:96:in `create'
will_paginate (3.1.0) lib/will_paginate/active_record.rb:125:in `to_a'
activerecord (4.1.14.1) lib/active_record/relation.rb:602:in `inspect'
activerecord (4.1.14.1) lib/active_record/relation/delegation.rb:46:in `rescue in each'
activerecord (4.1.14.1) lib/active_record/relation/delegation.rb:46:in `each'
app/controllers/recurring_events_controller.rb:20:in `index'
actionpack (4.1.14.1) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
actionpack (4.1.14.1) lib/abstract_controller/base.rb:189:in `process_action'
actionpack (4.1.14.1) lib/action_controller/metal/rendering.rb:10:in `process_action'
actionpack (4.1.14.1) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:113:in `call'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:113:in `call'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:552:in `block (2 levels) in compile'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:502:in `call'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:502:in `call'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:86:in `run_callbacks'
actionpack (4.1.14.1) lib/abstract_controller/callbacks.rb:19:in `process_action'
actionpack (4.1.14.1) lib/action_controller/metal/rescue.rb:29:in `process_action'
actionpack (4.1.14.1) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
activesupport (4.1.14.1) lib/active_support/notifications.rb:159:in `block in instrument'
activesupport (4.1.14.1) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
activesupport (4.1.14.1) lib/active_support/notifications.rb:159:in `instrument'
actionpack (4.1.14.1) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
actionpack (4.1.14.1) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
activerecord (4.1.14.1) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
actionpack (4.1.14.1) lib/abstract_controller/base.rb:136:in `process'
actionview (4.1.14.1) lib/action_view/rendering.rb:30:in `process'
actionpack (4.1.14.1) lib/action_controller/metal.rb:196:in `dispatch'
actionpack (4.1.14.1) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
actionpack (4.1.14.1) lib/action_controller/metal.rb:232:in `block in action'
actionpack (4.1.14.1) lib/action_dispatch/routing/route_set.rb:80:in `call'
actionpack (4.1.14.1) lib/action_dispatch/routing/route_set.rb:80:in `dispatch'
actionpack (4.1.14.1) lib/action_dispatch/routing/route_set.rb:48:in `call'
actionpack (4.1.14.1) lib/action_dispatch/journey/router.rb:73:in `block in call'
actionpack (4.1.14.1) lib/action_dispatch/journey/router.rb:59:in `each'
actionpack (4.1.14.1) lib/action_dispatch/journey/router.rb:59:in `call'
actionpack (4.1.14.1) lib/action_dispatch/routing/route_set.rb:690:in `call'
warden (1.2.6) lib/warden/manager.rb:35:in `block in call'
warden (1.2.6) lib/warden/manager.rb:34:in `catch'
warden (1.2.6) lib/warden/manager.rb:34:in `call'
rack (1.5.5) lib/rack/etag.rb:23:in `call'
rack (1.5.5) lib/rack/conditionalget.rb:25:in `call'
rack (1.5.5) lib/rack/head.rb:11:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/flash.rb:254:in `call'
rack (1.5.5) lib/rack/session/abstract/id.rb:225:in `context'
rack (1.5.5) lib/rack/session/abstract/id.rb:220:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/cookies.rb:562:in `call'
activerecord (4.1.14.1) lib/active_record/query_cache.rb:36:in `call'
activerecord (4.1.14.1) lib/active_record/connection_adapters/abstract/connection_pool.rb:621:in `call'
activerecord (4.1.14.1) lib/active_record/migration.rb:380:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
activesupport (4.1.14.1) lib/active_support/callbacks.rb:82:in `run_callbacks'
actionpack (4.1.14.1) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/reloader.rb:73:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/remote_ip.rb:76:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
railties (4.1.14.1) lib/rails/rack/logger.rb:38:in `call_app'
railties (4.1.14.1) lib/rails/rack/logger.rb:20:in `block in call'
activesupport (4.1.14.1) lib/active_support/tagged_logging.rb:68:in `block in tagged'
activesupport (4.1.14.1) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (4.1.14.1) lib/active_support/tagged_logging.rb:68:in `tagged'
railties (4.1.14.1) lib/rails/rack/logger.rb:20:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/request_id.rb:21:in `call'
rack (1.5.5) lib/rack/methodoverride.rb:21:in `call'
rack (1.5.5) lib/rack/runtime.rb:17:in `call'
activesupport (4.1.14.1) lib/active_support/cache/strategy/local_cache_middleware.rb:26:in `call'
rack (1.5.5) lib/rack/lock.rb:17:in `call'
actionpack (4.1.14.1) lib/action_dispatch/middleware/static.rb:84:in `call'
rack (1.5.5) lib/rack/sendfile.rb:112:in `call'
railties (4.1.14.1) lib/rails/engine.rb:514:in `call'
railties (4.1.14.1) lib/rails/application.rb:144:in `call'
rack (1.5.5) lib/rack/content_length.rb:14:in `call'
thin (1.6.4) lib/thin/connection.rb:86:in `block in pre_process'
thin (1.6.4) lib/thin/connection.rb:84:in `catch'
thin (1.6.4) lib/thin/connection.rb:84:in `pre_process'
thin (1.6.4) lib/thin/connection.rb:53:in `process'
thin (1.6.4) lib/thin/connection.rb:39:in `receive_data'
eventmachine (1.0.9.1) lib/eventmachine.rb:193:in `run_machine'
eventmachine (1.0.9.1) lib/eventmachine.rb:193:in `run'
thin (1.6.4) lib/thin/backends/base.rb:73:in `start'
thin (1.6.4) lib/thin/server.rb:162:in `start'
rack (1.5.5) lib/rack/handler/thin.rb:16:in `run'
rack (1.5.5) lib/rack/server.rb:264:in `start'
railties (4.1.14.1) lib/rails/commands/server.rb:69:in `start'
railties (4.1.14.1) lib/rails/commands/commands_tasks.rb:81:in `block in server'
railties (4.1.14.1) lib/rails/commands/commands_tasks.rb:76:in `tap'
railties (4.1.14.1) lib/rails/commands/commands_tasks.rb:76:in `server'
railties (4.1.14.1) lib/rails/commands/commands_tasks.rb:40:in `run_command!'
railties (4.1.14.1) lib/rails/commands.rb:17:in `<top (required)>'
bin/rails:4:in `require'
bin/rails:4:in `<top (required)>'
ruby-debug-ide (0.6.0) lib/ruby-debug-ide.rb:88:in `debug_load'
ruby-debug-ide (0.6.0) lib/ruby-debug-ide.rb:88:in `debug_program'
ruby-debug-ide (0.6.0) bin/rdebug-ide:130:in `<top (required)>'
-e:1:in `load'
-e:1:in `<main>'
```

### System configuration
**Rails version**: 4.1.14.1

**Ruby version**: 2.1.5p273

**Comment by rafaelfranca** [comment]
Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by brandonpancost** [comment]
I just wrote a script real quick. It should test the issue, but it needs to be switched from sqlite to mysql in order to reproduce the error.

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', github: 'rails/rails'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :hashtags, force: true  do |t|
    t.string :tag
    t.string :something
  end
  create_table :posts, force: true  do |t|
    t.string :tag
  end
end

class Hashtag < ActiveRecord::Base
  has_many :posts, primary_key: 'tag', foreign_key: 'tag'
end

class Post < ActiveRecord::Base
  belongs_to :hashtag, foreign_key: 'tag'
end

class BugTest < Minitest::Test
  def test_preload_accented_characters
    Hashtag.create(tag: 'normal', something: 'test')
    Post.create(tag: 'normal')

    Hashtag.create(tag: 'el_niño', something: 'test')
    Post.create(tag: 'el_nino')

    regular = Hashtag.where(something: 'test')
    assert_equal 2, regular.length
    regular.each do |row|
      assert_equal 1, row.posts.length
    end

    preloaded = Hashtag.preload(:posts).where(something: 'test')
    assert_equal 2, preloaded.length
    preloaded.each do |row|
      assert_equal 1, row.posts.length
    end
  end
end

```

**Comment by sevaorlov** [comment]
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader/association.rb#L67
records: {"normal"=>[#<Post id: 1, tag: "normal">], "el_nino"=>[#<Post id: 2, tag: "el_nino">]}
convert_key(owner[owner_key_name]): el_niño

it is actually hard to know which Post belong to which Hashtag after posts are loaded
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader/association.rb#L111

**Comment by sevaorlov** [comment]
we can use `I18n.transliterate` or something else to convert string with accented characters, but that implementation can diverge with mysql implementation and will need to be dependent from mysql column encoding settings. not sure that it is worth it.

**Comment by utilum** [comment]
I see a failure with both sqlite and mysql.

**Comment by al2o3cr** [comment]
I believe this is essentially the same issue as described here: https://github.com/rails/rails/issues/21668

TL;DR preloading associations joined with columns that don't obey strict equality (`citext` in the linked issue, transliterated fields here) is probably not a supported feature.

**Comment by vipulnsward** [comment]
Closing in favor of https://github.com/rails/rails/issues/24364
